### PR TITLE
fix #418 Use of inline styles doesn't play nice with Content Security Policy

### LINKED ISSUE #418: Use of inline styles doesn't play nice with Content Security Policy
gridstack.js seems to use inline styles extensively. We would like to disable inline styles via CSP's style-src to avoid the security risk they present. Are there any plans for gridstack.js to avoid the use of inline-styles?

### PR DESCRIPTION
**Pull request by shaurya-sisodia**
fix #418
### Description
* updating styles directly on HTMLELement style description rather than updating dynamic stylesheets
* removing logic related to dynamic stylesheets as they aren't required and violate CSP strict style-src rule
* removed UTs related to Dynamic stylesheets

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [x] All tests passing (`yarn test`)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: src/gridstack.ts [hunk 1]
```diff
@@ -226,8 +221,8 @@ export class GridStack {
   protected _ignoreLayoutsNodeChange: boolean;
   /** @internal */
   public _gsEventHandler = {};
-  /** @internal */
-  protected _styles: GridCSSStyleSheet;
+  /** @internal max height of the gridstack*/
+  public _max: number;
```
Comment (adumesny) [inline src/gridstack.ts:225]: does it need to be public ? if so '_' isn't the right prefix, and rename to curRow or something (need to see how it's used)

File: src/gridstack.ts [hunk 2]
```diff
@@ -240,6 +235,8 @@ export class GridStack {
   protected _extraDragRow = 0;
   /** @internal true if nested grid should get column count from our width */
   protected _autoColumn?: boolean;
+  /** @internal calculates height for a given row */
+  private _getHeight: (row: number) => string;
```
Comment (adumesny) [inline src/gridstack.ts:239]: make this a proper TS method, not var you assign to function later.

File: src/gridstack.ts [hunk 3]
```diff
@@ -338,6 +335,8 @@ export class GridStack {
 
     this._setStaticClass();
 
+    this._getHeight = (rows: number) => (rows * (this.opts.cellHeight as number)) + this.opts.cellHeightUnit;
```
Comment (adumesny) [inline src/gridstack.ts:338]: see above

File: src/gridstack.ts [hunk 4]
```diff
@@ -1209,47 +1198,38 @@ export class GridStack {
...
-    if (!this._styles) {
-      let id = 'gridstack-style-' + (Math.random() * 100000).toFixed();
-      // insert style to parent (instead of 'head' by default) to support WebComponent
-      let styleLocation = this.opts.styleInHead ? undefined : this.el.parentNode as HTMLElement;
-      this._styles = Utils.createStylesheet(id, styleLocation);
-      if (!this._styles) return this;
-      this._styles._id = id;
-      this._styles._max = 0;
+    if (!this._max) {
+      this._max = 0;
```
Comment (adumesny) [inline src/gridstack.ts:1202]: doesn't prevent if (!this._max) {...} clause, so likely incorrect placement.

File: src/gridstack.ts [hunk 5]
```diff
@@ -1209,47 +1198,38 @@ export class GridStack {
...
+      Utils.updateStyleOnElements(`${prefix} > .ui-resizable-e`, {right});
+      Utils.updateStyleOnElements(`${prefix} > .ui-resizable-se`, {right, bottom});
+      Utils.updateStyleOnElements(`${prefix} > .ui-resizable-nw`, {left});
+      Utils.updateStyleOnElements(`${prefix} > .ui-resizable-w`, {left});
+      Utils.updateStyleOnElements(`${prefix} > .ui-resizable-sw`, {left, bottom});
     }
 
+    // apply layout styles on all widgets
+    const widgets: HTMLElement[] = Utils.getElements(prefix);
+    for (const widget of widgets) {
```
Comment (adumesny) [inline src/gridstack.ts:1226]: Utils.getElements(prefix).forEach(w => Utils.updatePositionStyleOnWidget(w, this._getHeight));

File: src/utils.ts [hunk 6]
```diff
@@ -3,404 +3,401 @@
...
-        ret[key] = Utils.cloneDeep(obj[key]);
-      }
-    }
-    return ret;
-  }
-}
-
-// list of fields we will skip during cloneDeep (nested objects, other internal)
-const skipFields = ['_isNested', 'el', 'grid', 'subGrid', 'engine'];
+ import { GridStackElement, GridStackNode, GridStackOptions, numberOrString, GridStackPosition, GridStackWidget } from './types';
```
Comment (adumesny) [inline src/utils.ts:6]: indent all left so we can see actual changes...

**Review by adumesny — COMMENTED** [review]
thank you for looking into this. Have  a few comments and I will need to pull it locally later to test as well.

**Comment by shaurya-sisodia** [comment]
Hi @adumesny 

Thank you so much for such quick feedback.

I have incorporated the review comments. I have a couple of questions. I have fixed the wrong indent that I committed.
Please have a look. 

Regards,
Shaurya

File: src/gridstack.ts [hunk 7]
```diff
@@ -1305,6 +1280,7 @@ export class GridStack {
     if (n.y !== undefined && n.y !== null) { el.setAttribute('gs-y', String(n.y)); }
     if (n.w) { el.setAttribute('gs-w', String(n.w)); }
     if (n.h) { el.setAttribute('gs-h', String(n.h)); }
+    Utils.updatePositionStyleOnWidget(el, this._getHeight);
```
Comment (adumesny) [inline src/gridstack.ts:1283]: I would pass n as that is the source of truth, not the gs-*** attributes

File: src/utils.ts [hunk 8]
```diff
@@ -114,46 +114,42 @@ export class Utils {
...
-    } else if (typeof sheet.insertRule === 'function') {
-      sheet.insertRule(`${selector}{${rules}}`);
-    }
+  /** update position style on element */
+  static updatePositionStyleOnWidget(el: HTMLElement, getHeight:(rows: number) => string): void {
+    const h = Utils.toNumber(el.getAttribute('gs-h'));
+    const minH= Utils.toNumber(el.getAttribute('gs-min-h'));
+    const maxH = Utils.toNumber(el.getAttribute('gs-max-h'));
+    const y = Utils.toNumber(el.getAttribute('gs-y'));
+    el.style.height = getHeight(h);
```
Comment (adumesny) [inline src/utils.ts:149]: I would pass GridStackPosition instead.
and you are setting the height(s) and top only - what about left and widths ? seem one sided...

File: src/utils.ts [hunk 9]
```diff
@@ -114,46 +114,42 @@ export class Utils {
...
-    }
-    if (!parent) {
-      // default to head
-      parent = document.getElementsByTagName('head')[0];
-      parent.appendChild(style);
-    } else {
-      parent.insertBefore(style, parent.firstChild);
+  /** update CSS style on elements with given selector */
+  static updateStyleOnElements(selector: string, styles: {[props: string]: string |  string[] }): void {
+    const elements = Utils.getElements(selector);
```
Comment (adumesny) [inline src/utils.ts:119]: Utils.getElements(selector).forEach(el => Utils.addElementStyle(el, styles));

File: src/utils.ts (hunk 8 above)
Comment (adumesny) [reply src/utils.ts:149]: also I would not set properties (like min/max) that are not set

File: src/utils.ts [hunk 10]
```diff
@@ -114,46 +114,42 @@ export class Utils {
...
     }
-    return style.sheet as CSSStyleSheet;
   }
 
-  /** removed the given stylesheet id */
-  static removeStylesheet(id: string): void {
-    let el = document.querySelector('STYLE[gs-style-id=' + id + ']');
-    if (el && el.parentNode) el.remove();
+  /** update CSS style on element */
+  static addElementStyle(el: HTMLElement, styles: {[props: string]: string |  string[] }): void {
```
Comment (adumesny) [inline src/utils.ts:126]: if (!(styles instanceof Object)) return; // not sure this even needed if (!styles) is enough
for (const key in styles) {
  if (!styles.hasOwnProperty(key)) return;
  const style = styles[key];
  if (Array.isArray(style)) { // why do we need this ?
    style.forEach(val => el.style[key] = val);
  } else {
    el.style[key] = style;
  }
}

**Review by adumesny — COMMENTED** [review]
also your code will update all the elements styles when one of them changes, which will be very inefficient (prev code only did incremental stylesheet update as maxRow was reached).

without deleting gridstack.css (or at least any widget location/size info) and gridstack-extra.css you can't tell if you have handled all cases (I doubt it). Are you ONLY removing JS code that create inline styles, but still allowing static styles sheets (with dynamic class setting). I don't know enough about CSP to tell how far we would need to go...

this will need very different architecture.

**Comment by shaurya-sisodia** [comment]
Hi @adumesny 

CSP strict style rule prohibits the following:

- Inserting rule using CSSStyleSheet.insertRule() or CSSStyleSheet.addRule()
- Appending the an element with inline style in the dom: Eg- El.appendChild(`<div style="width:100px"></div>`);
- Appending dynamically generated CSSStyleSheets into the DOM.

That's why I had to change the implementation such that we set styles on each element. 
In my architecture, we have to update Styles on each element because we have to take into account that the position on multiple elements might have changed too besides the CSS style calculation itself.

We can use the old architecture and still make it CSP compliant for all browsers except SAFARI. 
As written here in the following issue: 
I have come across [replaceSync accepts inline CSS which is not CSP compliant · Issue #98 · WICG/construct-stylesheets (github.com)](https://github.com/WICG/construct-stylesheets/issues/98)

We can use CSSStyleSheet.replaceSync() instead of insertRule() to update styles on dynamic style sheet. We can also instead of appending this dynamic stylesheet can use
`document.adoptedStyleSheets = document.adoptedStyleSheets.concat(sheet);`
 
 
 We can use old architecture with adoptedStyleSheets and replaceSync is all the browsers and new architecture in case of SAFARI. Or ask applications to use precalculated stylesheets in case of safari similar to [IE 8 Support](https://github.com/iongroup/gridstack.js/blob/v0.2.5-branch/README.md#ie8-support).

**Comment by adumesny** [comment]
so we have a mix of css rules + classes for thing that don't change (eg column based layout) but attributes for height stuff...nthat would satisfy CSP but looks odd. Also we do NOT need to update all elements as we know when a node is dirty and needs to be updated (we already track that) so the call to generate styles need to taken out as it was done only of we added rows (or cellHeight/margin changed). Your logic isn't correct.

also keep in m id you changed break some of the demos so there are issues there - just playing with two.html shows issues dragging/resizing around. https://gridstackjs.com/demo/two.html is fine while your version using demo/two.html is not

**Comment by shaurya-sisodia** [comment]
Hi @adumesny 

I am aware of the demos breaking and have fixed the changes locally and will commit shortly.

Also so we can use the architecture but update styles on elements only when they are dirty?

**Comment by shaurya-sisodia** [comment]
Hi @adumesny 

I have fixed the code and incorporated the review you sent. The demo is also working perfectly.  We are updating styles on dirty ndes only

**Comment by shaurya-sisodia** [comment]
@adumesny  Please review

**Comment by adumesny** [comment]
this is likely not needed (and quite a risky change) given `nonce` support we now have.
https://github.com/gridstack/gridstack.js/pull/2229

I would like to close it - we can always get back to it from 'won't fix' issue listed.

**Pull request closed without merging**